common/djangoapps/track/utils.py

```diff
@@ -28,3 +30,38 @@ def default(self, obj):  # pylint: disable=method-hidden
             return obj.isoformat()
 
         return super(DateTimeJSONEncoder, self).default(obj)
+
+
+class LatinJSONEncoder(DateTimeJSONEncoder):
+    """
+    JSON encoder for UnicodeDecodeError with latin1 characters
+    """
+
+    def encode(self, obj):
+        """
+        encode method gets an original object
+        and returns result string. obj argument will be the
+        object that is passed to json.dumps function
+        """
+        obj = self.iterate_dictionary(obj)
+
+        return super(LatinJSONEncoder, self).encode(obj)
+
+    def iterate_dictionary(self, obj):
+        """
+        Iterate over nested dict structure
+        """
+        for key, value in obj.iteritems():
+            if isinstance(value, dict):
+                self.iterate_dictionary(value)
+            elif isinstance(value, str):
```

Comment:
Thanks, Adam.  Finding out why this was happening is also my supreme goal here.  If we want to just add extra logging, that can be done at the point that the event is lost, without trying to hack through a way to "fix" the bad data in a way that doesn't break.

My concern is that as I explore the kinds of data that gets dumped for events, this effort to "fix" the data will end up failing anyway.  The structure of event data is increasingly nested, and it includes various combinations of dicts and lists, as well as strings and ints and such.  There are definitely dicts that contain lists as well as strings and other dicts, and lists that contain dicts as well as strings.  (I'm not sure I've seen lists that contain lists.)  I've been doing a cursory inventory for the purposes of deidentification, and have been surprised by some of what gets written.